x/sync/network_server.go

```diff
@@ -136,32 +141,100 @@ func (s *NetworkServer) HandleChangeProofRequest(
+	for keyLimit > 0 {
+		changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(keyLimit))
+		if err != nil {
+			// handle expected errors so clients cannot cause servers to spam warning logs.
+			if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
+				s.log.Debug(
+					"dropping invalid change proof request",
+					zap.Stringer("nodeID", nodeID),
+					zap.Uint32("requestID", requestID),
+					zap.Stringer("req", req),
+					zap.Error(err),
+				)
+				return nil // dropping request
+			}
+			return err
+		}
+
+		proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
+		if err != nil {
+			return err
+		}
+		if len(proofBytes) < bytesLimit {
+			return s.appSender.SendAppResponse(ctx, nodeID, requestID, proofBytes)
+		}
+		// the proof size was too large, try to shrink it
 
-	changeProof, err := s.db.GetChangeProof(ctx, req.StartingRoot, req.EndingRoot, req.Start, req.End, int(limit))
-	if err != nil {
-		// handle expected errors so clients cannot cause servers to spam warning logs.
-		if errors.Is(err, merkledb.ErrRootIDNotPresent) || errors.Is(err, merkledb.ErrStartRootNotFound) {
-			s.log.Debug(
-				"dropping invalid change proof request",
+		// ensure that the new limit is always smaller
+		keyLimit = uint16((len(changeProof.KeyValues) + len(changeProof.DeletedKeys)) / 2)
+
+		// estimate the bytes of the start and end proof to ensure that everything will fit into the bytesLimit
+		bytesEstimate := getBytesEstimateOfProofNodes(changeProof.StartProof)
+
+		// just the start proof is too large, so a proof is impossible
+		if bytesEstimate > int(req.BytesLimit) {
+			// errors are fatal, so log for the moment
+			s.log.Warn(
+				"cannot generate a proof within bytes limit",
 				zap.Stringer("nodeID", nodeID),
 				zap.Uint32("requestID", requestID),
 				zap.Stringer("req", req),
-				zap.Error(err),
+				zap.Error(ErrMinProofSizeIsTooLarge),
 			)
-			return nil // dropping request
+			return nil
 		}
-		return err
-	}
 
-	proofBytes, err := merkledb.Codec.EncodeChangeProof(Version, changeProof)
-	if err != nil {
-		return err
+		bytesEstimate += getBytesEstimateOfProofNodes(changeProof.EndProof)
+		deleteKeyIndex := 0
+		changeKeyIndex := 0
+
+		// shrink more if the early keys are extremely large
+		for keyIndex := uint16(1); keyIndex < keyLimit; keyIndex++ {
+			// determine if the deleted key or changed key is the next smallest key
+			currentBytes := 0
+
+			// if there is a deleted key at deleteKeyIndex and
+			// (there are no more change keys or the changed key is larger than the deleted key)
+			if deleteKeyIndex < len(changeProof.DeletedKeys) &&
+				(changeKeyIndex >= len(changeProof.KeyValues) ||
+					bytes.Compare(changeProof.KeyValues[changeKeyIndex].Key, changeProof.DeletedKeys[deleteKeyIndex]) > 0) {
+				currentBytes = len(changeProof.DeletedKeys[deleteKeyIndex]) + binary.MaxVarintLen64
```

Comment:
Instead of using `binary.MaxVarintLen64` can we export something from the codec so we're not assuming stuff about its implementation details here?